File: .github/scripts/append-json.js
const fs = require("fs");
const path = require("path");
const axios = require("axios");

const invalid_chars = [
  ":", "/", "\'", "\"", "?", "#", "%", "&", "{", "}", "\\", "<", ">", "*", "$",
  "!", "@", "+", "`", "|", "=", "."
];

function filterFilename(name) {
  invalid_chars.forEach((c) => {
    name = name.split("").filter((i) => i !== c).join("");
  });
  return name;
}

async function run() {
  try {
    // Get the PR number
    const prNumber = process.env.PR_NUMBER;
    const repo = process.env.GITHUB_REPOSITORY;
    const token = process.env.GITHUB_TOKEN;
    // Get the repository owner and name
    console.log(`Fetching details for PR ${prNumber} in repository ${repo}`);

    // Get the PR details using the GitHub API
    const response = await axios.get(`https://api.github.com/repos/${repo}/pulls/${prNumber}`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    const response_files = await axios.get(`https://api.github.com/repos/${repo}/pulls/${prNumber}/files`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    console.log(response_files.data);
    // Set file variables
    let bin_file = null;
    let midi_file = null;
    let preview_file = null;
    let preview_extension = null;
    const preview_extensions = ["wav", "mp3"];
    for (i = 0; i < response_files.data.length; i++) {
      const f = response_files.data[i];
      console.log(f)
      const extension_sep = f.filename.split(".");
      const extension = extension_sep[extension_sep.length - 1];
      if (extension == "bin") {
        bin_file = f.filename;
      } else if (extension == "mid") {
        midi_file = f.filename;
      } else if (preview_extensions.includes(extension)) {
        preview_file = f.filename;
        preview_extension = extension;
      }
    }
    console.log(bin_file, midi_file, preview_file, preview_extension)

    // Extract the PR message
    const prMessage = response.data.body;
    const rawPRData = prMessage.split("\r\n")
    const REQ_STRING = "IS SONG - DO NOT DELETE THIS LINE"
    if (rawPRData[0] != REQ_STRING) {
        console.log("Skipping Pull Request, missing key line.");
        return;
    }
    let json_output = {}
    rawPRData.forEach((item, index) => {
        if (index > 0) {
            if (item.split("").includes(":")) {
              spl = item.split(/:(.*)/s)
              json_output[spl[0].trim()] = spl[1].trim()
            }
        }
    })
    console.log(json_output)
    const number_vars = ["Tracks", "Duration"]
    number_vars.forEach((v) => {
        if (Object.keys(json_output).includes(v)) {
            if (!isNaN(json_output[v])) {
                json_output[v] = Number(json_output[v])
            }
        }
    })
    const arr_vars = ["Categories"]
    arr_vars.forEach((v) => {
        if (Object.keys(json_output).includes(v)) {
            json_output[v] = json_output[v].split(",").map((item) => item.trim())
        }
    })
    json_output["Verified"] = true;
    dt = new Date();
    json_output["Date"] = dt.toString();
    // Read the existing JSON file
    console.log(__dirname)
    const file = "mapping.json"
    const filePath = path.join(__dirname, `../../${file}`);
    const existingData = fs.existsSync(filePath) ? require(filePath) : [];
    // Get new file name
    let revisions = existingData.filter((entry) => ((entry["Game"] == json_output["Game"]) && (entry["Song"] == json_output["Song"]))).length;
    const rev_string = revisions == 0 ? "" : ` (REV ${revisions})`;
    const sub_file = `${filterFilename(json_output["Game"])}/${filterFilename(json_output["Song"])}${rev_string}`
    let binary_link = null;
    if (preview_file) {
      json_output["Audio"] = encodeURI(`https://github.com/theballaam96/candys-shop/raw/main/previews/${sub_file}.${preview_extension}`)
    }
    if (bin_file) {
      json_output["Binary"] = `binaries/${sub_file}.bin`
      binary_link = encodeURI(`https://github.com/theballaam96/candys-shop/raw/main/binaries/${sub_file}.bin`)
    }
    console.log(sub_file)

    
    // Read existing file for PR
    console.log("Starting file transfer")
    file_data = {
      "binaries": [bin_file, "bin", true],
      "previews": [preview_file, preview_extension, true],
      "midi": [midi_file, "mid", false],
    }
    let preview_file_bytes = null;
    Object.keys(file_data).forEach((k) => {
      k_file = file_data[k][0];
      k_ext = file_data[k][1];
      k_keep = file_data[k][2];
      if (k_file != null) {
        const binFilePath = path.join(__dirname, `../../${k_file}`);
        if (k_keep) {
          const newBinFile = `${k}/${sub_file}.${k_ext}`
          const rootDir = `${k}`
          const gameDir = `${k}/${filterFilename(json_output["Game"])}`
          if (!fs.existsSync(rootDir)) {
            fs.mkdirSync(rootDir)
          }
          if (!fs.existsSync(gameDir)) {
            fs.mkdirSync(gameDir)
          }
          const binNewFilePath = path.join(__dirname, `../../${newBinFile}`);
          const binFileData = fs.existsSync(binFilePath) ? fs.readFileSync(binFilePath) : null;
          if (k == "previews") {
            preview_file_bytes = binFileData;
          }
          if (binFileData != null) {
            fs.writeFileSync(binNewFilePath, binFileData);
          }
        }
        fs.unlinkSync(binFilePath);
      }
    })
    console.log("File Transfer Done")

    // Append the PR message to the JSON file
    existingData.push(json_output);

    // Write the updated JSON file
    fs.writeFileSync(filePath, JSON.stringify(existingData, null, 2));

    console.log("PR message appended to JSON file successfully.");

    // Post to discord
    const shorthand_to_long = {
      "bgm": "BGM",
      "events": "Event",
      "majoritems": "Major Item",
      "minoritems": "Minor Item"
    };
    let suggested_type = "Not Provided";
    if (Object.keys(json_output).includes("Group")) {
      suggested_type = json_output["Group"];
      if (Object.keys(shorthand_to_long).includes(suggested_type)) {
        suggested_type = shorthand_to_long[json_output["Group"]];
      }
    }
    const song_name = Object.keys(json_output).includes("Song") ? json_output["Song"] : "Not Provided"
    let desc_information = {
      "Game": Object.keys(json_output).includes("Game") ? json_output["Game"] : "Not Provided",
      "Song Name": Object.keys(json_output).includes("Song") ? json_output["Song"] : "Not Provided",
      "Type": suggested_type,
      "Tags": Object.keys(json_output).includes("Categories") ? json_output["Categories"] : "Not Provided",
    }
    if (Object.keys(json_output).includes("Update Notes")) {
      desc_information["Additional Notes"] = json_output["Update Notes"];
    }
    let desc_strings = []
    Object.keys(desc_information).forEach(k => {
      desc_strings.push(`**${k}**: ${desc_information[k]}`)
    })
    let update_string = ""
    if (revisions > 0) {
      update_string = ":watermelon: This song is an update\n"
    }
    let desc = desc_strings.join("\n")
    const binary_dl_link = binary_link ? binary_link : "???";
    const has_audio_file = preview_file_bytes != null;
    let audio_string = "No Preview";
    if (has_audio_file) {
      audio_string = "*(Attached)*";
    } else if (Object.keys(json_output).includes("Audio")) {
      audio_string = json_output["Audio"];
    }
    const embeds_arr = [{
      title: song_name,
      description: `${update_string}${desc}`,
      fields: [
        {
          name: "Composer(s)",
          value: Object.keys(json_output).includes("Composers") ? json_output["Composers"] : "Not Provided",
          inline: true,
        },
        {
          name: "Converted By",
          value: Object.keys(json_output).includes("Converters") ? json_output["Converters"] : "Not Provided",
          inline: true,
        },
        {
          name: "",
          value: `Download: ${binary_dl_link}`
        },
        {
          name: "",
          value: `Listen: ${audio_string}`
        }
      ],
      timestamp: new Date().toISOString(),
    }]

    const webhookUrl = process.env.DISCORD_WEBHOOK_PUBLICFILE;
    const options = {
        method: "POST",
        url: webhookUrl,
        headers: { "Content-Type": "application/json" },
        data: {
            content: "A new song has been uploaded",
            embeds: embeds_arr,
        },
    }
    axios(options)
        .then(whresp => {
            console.log('Message sent successfully:', whresp.data);
            if (has_audio_file) {
              const metadata = {}
              const boundary = "xxxxxxxx"
              let data = "";
              for (let i in metadata) {
                data += `--${boundary}\r\n`;
                data += `Content-Disposition: form-data; name=\"${i}\"; \r\n\r\n${metadata[i]}\r\n`;
              }
              data += "--" + boundary + "\r\n";
              const new_song_name = `${song_name.replaceAll(' ','').replaceAll("\"","")}.${preview_extension}`
              data += "Content-Disposition: form-data; name=\"file\"; filename=\"" + new_song_name + "\"\r\n";
              data += "Content-Type:" + "audio/mpeg" + "\r\n\r\n";
              binary_buffer = Buffer.from(preview_file_bytes, "binary");
              let payload_data = Utilities.newBlob(data).getBytes()
                .concat(binary_buffer)
                .concat(binary_buffer);
              options = {
                method: "post",
                url: webhookUrl,
                headers: {
                  "Content-Type": "multipart/form-data; boundary=" + boundary
                },
                data: payload_data
              }
              axios(options).then(whresp2 => {
                console.log('Preview posted successfully:', whresp2.data);
              }).catch(error => {
                console.log(error.message);
                process.exit(1);
              });
            }
        })
        .catch(error => {
            console.log(error.message);
            process.exit(1);
        });
  } catch (error) {
    console.error("Error:", error.response ? error.response.data : error.message || error);
    process.exit(1);
  }
}

run();

File: .github/scripts/python/append_json.py
#!/usr/bin/env python3
"""
pr_to_mapping.py

Converted from the original Node.js script to Python for running in GitHub Actions.

Environment variables required:
- PR_NUMBER
- PAT_TOKEN
- DISCORD_WEBHOOK_PUBLICFILE

Dependencies (install in your action environment):
pip install requests mido

Notes:
- This script expects the uploaded files (bin/mid/wav/mp3) to be present at ../../<filename>
  relative to the script directory, matching the original Node.js behavior.
- mapping.json is expected (or will be created) at ../../mapping.json relative to script.
"""

import os
import sys
import json
from datetime import datetime
import requests
from dotenv import load_dotenv
from donk_lib import (
    compute_midi_duration_from_bytes,
    REPO,
    pull_pr_data,
    getHeaders,
    filter_filename,
    adjust_raw_url,
    getPath,
    postAudio,
)
from yt_post import uploadVideoWrapper

# ---------------------------
# Utility / configuration
# ---------------------------

load_dotenv()

# Paths: match the Node behavior of using __dirname and ../../
MAPPING_FILE = getPath("mapping.json")
MOVE_FILES = False
SHORTHAND_TO_LONGHAND = {
    "bgm": "BGM",
    "events": "Event",
    "majoritems": "Major Item",
    "minoritems": "Minor Item",
}


def message(json_output, binary_link, preview_file_bytes, is_update, preview_extension, preview_path):
    # Initial information
    webhook_url = os.getenv("DISCORD_WEBHOOK_PUBLICFILE")
    song_name = json_output.get("Song", "Not Provided")
    binary_dl_link = binary_link if binary_link else "???"
    has_audio_file = preview_file_bytes is not None
    update_string = ":watermelon: This song is an update\n" if is_update else ""
    if has_audio_file and (json_output["Category"] != "bgm"):
        # mimic Node Buffer length check: bytes length < (25 * 1024 * 1024)
        has_audio_file = len(preview_file_bytes) < (25 * 1024 * 1024)
    audio_string = "No Preview"
    game_name = json_output.get("Game", "Not Provided")
    composers = json_output.get("Composers", "Not Provided")
    converters = json_output.get("Converters", "Not Provided")
    add_yt = False
    if has_audio_file:
        if json_output["Category"] == "bgm":
            attempted_link = uploadVideoWrapper(game_name, song_name, converters, composers, preview_file_bytes)
            if attempted_link is not None:
                json_output["Audio"] = attempted_link
                audio_string = attempted_link
                if preview_path is not None and os.path.exists(preview_path):
                    os.remove(preview_path)
                add_yt = True
            else:
                audio_string = "*(Attached)*"
        else:
            audio_string = "*(Attached)*"
    elif "Audio" in json_output:
        audio_string = json_output["Audio"]
        add_yt = True
    suggested_type = "Not Provided"
    if "Category" in json_output:
        suggested_type = SHORTHAND_TO_LONGHAND.get(json_output["Category"], json_output["Category"])

    desc_information = {
        "Game": game_name,
        "Song Name": song_name,
        "Type": suggested_type,
        "Tags": ", ".join(json_output.get("Tags", ["Not Provided"])),
    }
    if "Update Notes" in json_output:
        desc_information["Additional Notes"] = json_output["Update Notes"]

    desc_strings = [f"**{k}**: {v}" for k, v in desc_information.items()]
    desc = '\n'.join(desc_strings)
    # Set up embeds
    embed = {
        "title": song_name,
        "description": f"{update_string}{desc}",
        "fields": [
            {"name": "Composer(s)", "value": composers, "inline": True},
            {"name": "Converted By", "value": converters, "inline": True},
        ],
        "timestamp": datetime.utcnow().isoformat(),
    }

    components = [
        {
            "type": 1,  # Action row
            "components": [
                {
                    "type": 2,  # Button
                    "style": 5,  # Link button
                    "label": "Binary File",
                    "url": binary_dl_link,
                },
            ],
        }
    ]
    if add_yt:
        components[0]["components"].append({
            "type": 2,
            "style": 5,
            "label": "YouTube",
            "url": audio_string,
        })

    payload = {
        "content": "A new song has been uploaded",
        "embeds": [embed],
        "components": components,  # Attach the buttons
    }

    # Post the embed
    try:
        whresp = requests.post(f"{webhook_url}?wait=true&with_components=true", json=payload, headers={"Content-Type": "application/json"})
        whresp.raise_for_status()
        print("Message sent successfully:", whresp.text)
    except Exception as e:
        print("Error sending webhook embed:", e)
        sys.exit(1)

    # If we have audio bytes and it is small enough, post as a multipart file
    if has_audio_file:
        postAudio(song_name, preview_extension, preview_file_bytes)

def handlePR(pr_number, check_labels = True):
    try:
        data = pull_pr_data(pr_number)
        if not data["is_song"]:
            return
        if "batch-merge-ignore" in data["labels"] and check_labels:
            return
        json_output = data["output"]

        # Read existing mapping.json
        if MAPPING_FILE.exists():
            try:
                with open(MAPPING_FILE, "r", encoding="utf-8") as fh:
                    existing_data = json.load(fh)
                if not isinstance(existing_data, list):
                    existing_data = []
            except Exception:
                existing_data = []
        else:
            existing_data = []

        # compute revisions
        revisions = 0
        for entry in existing_data:
            if (
                entry.get("Game") == json_output.get("Game")
                and entry.get("Song") == json_output.get("Song")
            ):
                revisions += 1
        rev_string = "" if revisions == 0 else f" (REV {revisions})"
        sub_file = f"{filter_filename(json_output.get('Game'))}/{filter_filename(json_output.get('Song'))}{rev_string}"

        binary_link = None
        if data["preview"]:
            json_output["Audio"] = requests.utils.requote_uri(
                f"https://github.com/{REPO}/raw/main/previews/{sub_file}.{data['preview_ext']}"
            )
        if data["bin"]:
            json_output["Binary"] = f"binaries/{sub_file}.bin"
            binary_link = requests.utils.requote_uri(
                f"https://github.com/{REPO}/raw/main/binaries/{sub_file}.bin"
            )
        print("sub_file:", sub_file)

        # File transfer (move/copy and cleanup)
        print("Starting file transfer")
        file_data = {
            "binaries": [data["bin"], "bin", True],
            "previews": [data["preview"], data["preview_ext"], True],
            "midi": [data["midi"], "mid", False],
        }

        preview_file_bytes = None
        preview_path = None

        for k, v in file_data.items():
            k_file, k_ext, k_keep = v
            if not k_file:
                continue
            bin_file_path = getPath(k_file)
            if k_keep:
                new_bin_file = f"{k}/{sub_file}.{k_ext}"
                root_dir = getPath(k)
                game_dir = getPath(k, filter_filename(json_output.get("Game") or ""))
                # ensure directories
                root_dir.mkdir(parents=True, exist_ok=True)
                game_dir.mkdir(parents=True, exist_ok=True)
                bin_new_file_path = getPath(new_bin_file)
                bin_file_data = None
                if bin_file_path.exists():
                    bin_file_data = bin_file_path.read_bytes()
                # If preview, keep bytes for Discord post
                if k == "previews":
                    preview_file_bytes = bin_file_data
                    preview_path = bin_new_file_path
                if bin_file_data is not None:
                    bin_new_file_path.parent.mkdir(parents=True, exist_ok=True)
                    bin_new_file_path.write_bytes(bin_file_data)
            # Remove original file if exists
            if bin_file_path.exists() and MOVE_FILES:
                try:
                    bin_file_path.unlink()
                except Exception as e:
                    print("Warning: could not remove original file:", bin_file_path, e)

        print("File Transfer Done")

        # Parse midi duration if midi_file provided
        if data["midi"] and data["midi_raw"]:
            midi_url = adjust_raw_url(data["midi_raw"])
            try:
                r = requests.get(midi_url, headers=getHeaders())
                r.raise_for_status()
                midi_bytes = r.content
                duration = compute_midi_duration_from_bytes(midi_bytes)
                if duration:
                    json_output["Duration"] = duration
            except Exception as e:
                print("Warning: could not fetch/parse midi:", e)


        # Prepare Discord embed payload
        message(json_output, binary_link, preview_file_bytes, revisions > 0, data["preview_ext"], preview_path)
        
        # Append and save mapping.json
        existing_data.append(json_output)
        try:
            MAPPING_FILE.parent.mkdir(parents=True, exist_ok=True)
            with open(MAPPING_FILE, "w", encoding="utf-8") as fh:
                json.dump(existing_data, fh, indent=2, ensure_ascii=False)
            print("PR message appended to JSON file successfully.")
        except Exception as e:
            print("Error writing mapping.json:", e)
            sys.exit(1)

    except requests.HTTPError as he:
        resp = he.response
        try:
            print("HTTP Error:", resp.status_code, resp.text)
        except Exception:
            print("HTTP Error:", he)
        sys.exit(1)
    except Exception as ex:
        print("Error:", ex)
        sys.exit(1)


if __name__ == "__main__":
    pr_number = os.getenv("PR_NUMBER")
    handlePR(pr_number)